README.md
[![Release](https://img.shields.io/github/release/Y-Lab/Y-VOD.svg)](https://github.com/Y-Lab/Y-VOD/releases)
[![Commits](https://img.shields.io/github/commits-since/Y-Lab/Y-VOD/latest.svg)](https://github.com/Y-Lab/Y-VOD/commits/master)
[![Last Commit](https://img.shields.io/github/last-commit/Y-Lab/Y-VOD.svg)](https://github.com/Y-Lab/Y-VOD/commits/master)

# Y-VOD
Y-English Local Area Network Video-On-Demand Solution


RELEASE-NOTES.md
## RELEASE NOTES

### Version 1.0.1 - February 15, 2019

**NEW**
- **Model** - Add `demo` attribute to video data model #1
- **Feature** - Add demo lessons #2
- **Model** - Create video collection data models #8
- **Model** - Add `restricted_permit` attribute to device data model #9
- **Feature** - Only permitted devices can access restricted videos #10
- **Feature** - Add VB demo lessons, Y-GRE demo lessons, and Y-GRE test reviews #11
- **Model** - Add device lesson type data model #13
- **Model** - Add `view_point` attribute to lesson type data model #19
- **Model** - Add `login_required` attribute to lesson type data model #20
- **Template** - Add template macro `placeholder_widget()` #21
- **Feature** - Adopt Y-System user migration API #30
- **Feature** - Adopt Y-System progress report API #31
- **Feature** - Adopt Y-System user authentication API #32
- **Feature** - Authenticate user via Y-System #37
- **Feature** - Email production server errors to system operator #39
- **Model** - Add `synchronized` attribute to punch data model #40
- **Model** - Add `progress_threshold` attribute to lesson data model #41
- **Feature** - Adopt Y-System lesson access API #42
- **Feature** - Add communication error alert to users on video page #61

**IMPROVED**
- **Model** - Change `demo` attribute of video data model to `restricted` #7
- **Model** - Remove user creation data model #14
- **Model** - Change `created_at` attribute of user data model to `imported_at` #15
- **Model** - Remove `restricted_permit` attribute from device data model #16
- **Model** - Remove `restricted` attribute from video data model #17
- **Model** - Remove `order` attribute from lesson data model #18
- **Feature** - Simplify user profile overview page contents #22
- **Template** - Adopt static assets hosted on static.y-english.cn #25
- **Form** - Rearrange form modules #27
- **View** - Rearrange view modules #28
- **Model** - Remove ID type data model #33
- **Model** - Remove gender data model #34
- **Model** - Remove `id_number`, `id_type_id` and `gender_id` attributes from user data model #35
- **Feature** - Remove Y-VOD token authentication feature #36

**FIXED**
- **Feature** - Playback rate changes are not taken into account in total video play time calculation #23
- **Feature** - Synchronized punches continue to synchronize #50
- **Feature** - Punch synchronization failure due to existing unsynchronized punches #54
- **Toolkit** - Exception on `get_mac_address_from_ip(ip_address)` #63
- **Feature** - Exception on sending video progress data when video is paused #65

### Version 1.0.0 - December 24, 2018

**IMPORTANT NOTE**
> Boom! Excited!


config.py
# -*- coding: utf-8 -*-

'''config.py'''

import os
import sys
from datetime import datetime


basedir = os.path.abspath(os.path.dirname(__file__))


class Config:
    '''Config'''

    # Directories
    BASE_DIR = basedir
    DATA_DIR = os.path.join(BASE_DIR, 'data')
    VIDEO_DIR = os.path.join(DATA_DIR, 'videos')
    CACHE_DIR = os.path.join(BASE_DIR, 'cache')
    HLS_DIR = os.path.join(CACHE_DIR, 'hls')

    # SSL
    SSL_DISABLE = True

    # CSRF
    WTF_CSRF_CHECK_DEFAULT = False

    # Security
    SECRET_KEY = os.getenv('YVOD_SECRET_KEY')
    AUTH_TOKEN_SECRET_KEY = os.getenv('YVOD_AUTH_TOKEN_SECRET_KEY')
    AUTH_REMEMBER_LOGIN = True

    # SQLAlchemy
    SQLALCHEMY_RECORD_QUERIES = True
    SQLALCHEMY_TRACK_MODIFICATIONS = False
    SLOW_DB_QUERY_TIME = 0.5 # seconds

    # Query
    RECORD_PER_PAGE = 20
    RECORD_PER_PAGE_FEWER = 10
    RECORD_PER_QUERY = 50

    # Time
    DATETIME_FORMAT = '%Y-%m-%dT%H:%M:%S'
    DATETIME_FORMAT_ISO = '%Y-%m-%dT%H:%M:%SZ'
    UTC_OFFSET = 8 # hours
    COOKIE_MAX_AGE = 30 * 24 * 60 * 60 # 30 days
    STATUS_EXPIRATION_CHECK_INTERVAL = 1000 # milliseconds (1 second)
    TOKEN_EXPIRATION = 3600 # seconds
    REQUEST_TIMEOUT = 15 # seconds

    # Video Analytics
    VIDEO_ANALYTICS_ACCELERATING_FACTOR = 1.25 # speedup
    VIDEO_ANALYTICS_GRANULARITY = 100 # milliseconds (0.1 seconds)
    VIDEO_ANALYTICS_UPDATE_PUNCH_INTERVAL = 15 * 1000 # milliseconds (15 seconds)
    VIDEO_ANALYTICS_STATUS_EXPIRATION = 300 # seconds (5 minutes)

    # Y-System
    YSYS_URI = os.getenv('YVOD_YSYS_URL')

    # Development
    SYSTEM_OPERATOR_NAME = os.getenv('YVOD_SYSTEM_OPERATOR_NAME')
    SERVER_SERIAL = os.getenv('YVOD_SERVER_SERIAL')
    SERVER_MAC_ADDRESS = os.getenv('YVOD_SERVER_MAC_ADDRESS')

    # Version
    PYTHON_VERSION = '{0.major}.{0.minor}.{0.micro}'.format(sys.version_info)
    VERSION = '1.0.1'
    ESTABLISHED_AT = '2018-08-05T00:00:00+08:00'
    UPDATED_AT = datetime.utcnow().strftime(DATETIME_FORMAT_ISO)
    RENDERED_AT = datetime.utcnow

    @staticmethod
    def init_app(app):
        '''Config.init_app(app)'''
        pass


class DevelopmentConfig(Config):
    '''DevelopmentConfig(Config)'''

    ENV = 'development'
    DEBUG = True

    # Database
    SQLALCHEMY_DATABASE_URI = 'sqlite:///{}'.format(os.path.join(basedir, 'yvod-dev.sqlite'))

    # HLS
    HLS_ENABLE = False


class ProductionConfig(Config):
    '''ProductionConfig(Config)'''

    ENV = 'production'

    # Database
    SQLALCHEMY_DATABASE_URI = os.getenv('YVOD_DATABASE_URL')

    # Mail
    MAIL_SERVER = 'smtp.exmail.qq.com'
    MAIL_PORT = 587
    MAIL_USE_TSL = True
    MAIL_USERNAME = os.getenv('YVOD_MAIL_USERNAME')
    MAIL_PASSWORD = os.getenv('YVOD_MAIL_PASSWORD')
    MAIL_SUBJECT_PREFIX = os.getenv('YVOD_MAIL_SUBJECT_PREFIX')
    MAIL_SENDER = os.getenv('YVOD_MAIL_SENDER')
    SYSTEM_OPERATOR_MAIL = os.getenv('YVOD_SYSTEM_OPERATOR_MAIL')

    # HLS
    HLS_ENABLE = True

    @classmethod
    def init_app(cls, app):
        '''ProductionConfig.init_app(cls, app)'''
        Config.init_app(app)
        # email errors to the system operator
        import logging
        from logging.handlers import SMTPHandler
        credentials = None
        secure = None
        if getattr(cls, 'MAIL_USERNAME', None) is not None:
            credentials = (cls.MAIL_USERNAME, cls.MAIL_PASSWORD)
            if getattr(cls, 'MAIL_USE_TSL', None):
                secure = ()
        mail_handler = SMTPHandler(
            mailhost=(cls.MAIL_SERVER, cls.MAIL_PORT),
            fromaddr=cls.MAIL_SENDER,
            toaddrs=[cls.SYSTEM_OPERATOR_MAIL],
            subject='{} Application Error'.format(cls.MAIL_SUBJECT_PREFIX),
            credentials=credentials,
            secure=secure
        )
        mail_handler.setLevel(logging.ERROR)
        app.logger.addHandler(mail_handler)


config = {
    'development': DevelopmentConfig,
    'production': ProductionConfig,

    'default': DevelopmentConfig,
}
